### PR TITLE
Proto XEP: "Best practices for password hashing and storage"

### PR DESCRIPTION
**Pull request by SamWhited**
This proto XEP provides best practices for password hashing and storage, tuning PBKDF2, and selecting SASL mechanisms. It will likely need expert review.

**Rendered:** https://blog.samwhited.com/xeps/inbox/password-storage.html

### CODE REVIEW

File: inbox/password-storage.xml [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+</section1>
+<section1 topic='Requirements' anchor='reqs'>
+  <p>
+    This document makes specific recommendations for best practices on the
+    public Jabber network for both clients and servers.
+    It does not attempt to address private networks or proprietary services
+    which may have different requirements, use cases, and security models.
+    These recommendations include the hashing and storage of memorized secrets
+    and other authenticators, authentication, and compatibility between clients
+    and servers with respect to authentiation.
```
Comment (lnjX) [inline inbox/password-storage.xml:66]:
```suggestion
    and servers with respect to authentication.
```

File: inbox/password-storage.xml [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+    </p>
+    <p>
+      If clients know that they will only ever authenticate using a mechanism
+      such as SCRAM where the original password is not needed (for example if
+      the mechanism has been pinned) they SHOULD store the SCRAM bits or the
+      hashed and salted password instead of the original password.
+      However, if backwards compatibility with servers that only support the
+      PLAIN mechanism or other mechanisms that require using the original
+      password is required, clients MAY choose to store the original password
+      so long as an appropriate keystore is used.
```
Comment (lnjX) [inline inbox/password-storage.xml:163]:
```suggestion
      as long as an appropriate keystore is used.
```

**Comment by lnjX** [comment]
###### Finally a XEP with nicely formatted XML :)

**Comment by dwd** [comment]
It's not clear from either this document or any of the references what "pepper" is, and while I've found the detail elsewhere eventually, this really needs clarification.

File: inbox/password-storage.xml (hunk 1 above)
Comment (SamWhited) [reply inbox/password-storage.xml:66]: Fixed, thanks!

File: inbox/password-storage.xml (hunk 2 above)
Comment (SamWhited) [reply inbox/password-storage.xml:163]: I think this one is okay, otherwise you wind up with "as long as" which sounds a bit odd to me (too many "as"'s)

**Comment by SamWhited** [comment]
> It's not clear from either this document or any of the references what "pepper" is, and while I've found the detail elsewhere eventually, this really needs clarification.

Fixed, thank you. The OWASP password cheat sheet one makes recommendations, but I guess it never defines it, oops!

**Pull request merged**